core/coretypes/include/coretypes/enumeration.h

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2022-2023 Blueberry d.o.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+#include <coretypes/common.h>
+#include <coretypes/stringobject.h>
+#include <coretypes/coretype.h>
+#include <coretypes/type_manager.h>
+#include <coretypes/enumeration_type.h>
+
+BEGIN_NAMESPACE_OPENDAQ
+
+/*!
+ * @ingroup types_enumeration
+ * @defgroup types_enumerations_enumeration Enumeration
+ * @{
+ */
+
+/*!
+ * @brief Enumerations are objects that encapsulate a value within a predefined set of named integral constants.
+ * These constants are predefined in an Enumeration type with the same name as the Enumeration.
+ *
+ * The Enumeration types are stored within a Type manager. In any given instance of openDAQ, a single Type manager should
+ * exist that is part of its Context.
+ *
+ * When creating an Enumeration object, the Type manager is used to validate the given enumerator value name against the
+ * Enumeration type stored within the manager. If no type with the given Enumeration name is currently stored,
+ * construction of the Enumeration object will fail. Similarly, if the provided enumerator value name is not part of
+ * the Enumeration type, the construction of the Enumeration object will also fail.
+ */
+DECLARE_OPENDAQ_INTERFACE(IEnumeration, IBaseObject)
+{
+    /*!
+     * @brief Gets the Enumeration's type.
+     * @param[out] type The Enumeration type
+     */
+    virtual ErrCode INTERFACE_FUNC getEnumerationType(IEnumerationType** type) = 0;
+
+    /*!
+     * @brief Gets the Enumeration value as String containing the name of the enumerator constant.
+     * @param[out] value Emumeration value.
+     */
+    virtual ErrCode INTERFACE_FUNC getValue(IString** value) = 0;
```

Comment:
Are enumerations mutable, or immutable objects - as in, can we select another value after they are created? Looking at the definition of core types they should not be mutable, but I'm not sure whether they're useful in that case.

Reply:
As was discussed enumeration object kept immutable and there is an [addition](https://github.com/openDAQ/openDAQ/pull/85/commits/b47563cbee24adcda2656b4cbb17fe74d90768bd#diff-bf70b119ff04f0d923254e2dc9f807b4e36d0179f1cad158b9673e04f2987143R116) to smart pointer implementation for assigning a new enumeration object created with the given value.